server/master/src/main/java/org/apache/accumulo/master/TimeoutTaskExecutor.java

```diff
@@ -0,0 +1,235 @@
+    this.exceptionCallback = Objects.requireNonNull(exceptionCallback, "Must provide a non-null exceptionCallback.");
+  }
+
+  /**
+   * Registers the callback to use on tasks that time out.
+   *
+   * @param timeoutCallback The callback function to invoke on timeouts.
+   * @throws NullPointerException when a null timeoutCallback is passed in
+   */
+  public void onTimeout(TimeoutCallback<C> timeoutCallback) {
+    this.timeoutCallback = Objects.requireNonNull(timeoutCallback, "Must provide a non-null timeoutCallback.");
+  }
+
+  /**
+   * Completes all the current tasks by dispatching to the appropriate callback.
+   *
+   * @throws IllegalStateException If all of the callbacks were not registered before calling this method.
+   * @throws InterruptedException  If interrupted while awaiting callable results.
+   */
+  public void complete() throws InterruptedException {
+    Preconditions.checkState(successCallback != null, "Must set a success callback before completing " + this);
+    Preconditions.checkState(exceptionCallback != null, "Must set an exception callback before completing " + this);
+    Preconditions.checkState(timeoutCallback != null, "Must set a timeout callback before completing " + this);
+
+    while (hasUnfinishedTasks()) {
+      completeTask(startedTasks.take());
+    }
+
+    wrappedTasks.clear();
+  }
+
+  private boolean hasUnfinishedTasks() {
+    for (WrappedTask wt : wrappedTasks) {
+      if (!wt.hasCompleted) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private void completeTask(WrappedTask wt) throws InterruptedException {
+    try {
+      long waitTime = wt.endTime - System.currentTimeMillis();
+      successCallback.accept(wt.callable, wt.future.get(waitTime, TimeUnit.MILLISECONDS));
+    } catch (InterruptedException e) {
+      throw e;
+    } catch (TimeoutException e) {
+      wt.future.cancel(true);
+      timeoutCallback.accept(wt.callable);
+    } catch (Exception e) {
+      exceptionCallback.accept(wt.callable, e);
+    }
+    wt.hasCompleted = true;
+  }
+
+  @Override
+  public void close() {
+    try {
+      executorService.shutdownNow();
+    } catch (Exception e) {
+      log.warn("Error while shutting down " + this, e);
+    }
+  }
+
+  /*
+   * A wrapper for a Callable that keeps additional information. This tracks the desired end time, if it has completed (either finished or cancelled), and keeps
+   * the associated future.
+   */
+  private class WrappedTask implements Callable<T> {
+    public final C callable;
+
+    // Set by worker thread and read by master thread
+    volatile long endTime;
+
+    // Set and read only by master thread
+    boolean hasCompleted = false;
+
+    public Future<T> future;
+
+    WrappedTask(C callable) {
+      this.callable = callable;
+    }
+
+    @Override
+    public T call() throws Exception {
+      startedTasks.put(this);
+      endTime = timeout + System.currentTimeMillis();
```

Comment:
Need to set `endTime` before adding self to startedTask.  Its possible that the task could be taken off the blocking queue before endTime is set.

Reply:
Nice catch. Thanks.